File: src/assets/nodes_builtin/and.py
"""AND 逻辑节点插件（type AND）：多组条件全部满足 → result=true。
条件结构与 selector 同构（branches[].condition.conditions[]，组内 logic 1=AND/2=OR），
逐组求值输出 results（每组 bool）+ result（AND 聚合）。配合 OR 节点做组合逻辑判定。
未设置的条件（编辑器占位）恒真——eval_condition_lenient（与批处理 filter 裁定一致）。
"""
from workflow_node_api import eval_condition_lenient


def _handle_and(node: dict, ctx) -> dict:
    branches = (node.get("data", {}).get("inputs", {}) or {}).get("branches") or []
    results = [bool(eval_condition_lenient((br.get("condition") or {}), ctx)) for br in branches]
    return {"outputs": {"result": all(results), "results": results}, "port": None}


def agt_node():
    return {"type": "AND", "label": "AND", "handler": _handle_and, "catalog": _CATALOG}


PARAMS = [
    {"key": "branches", "type": "object", "required": True,
     "desc": "条件组列表（与 selector 同构：每组含 logic 与 conditions），全部满足 → true"},
]

# ===== 节点目录条目（list_workflow_nodes / query_workflow_node 动态聚合自插件声明）=====
_CATALOG = {"name": "AND 逻辑与", "desc": "多组条件全部满足 → true。条件组与 selector 同构（左值/运算符/右值，支持引用上游字段），输出 result:boolean 与逐组结果列表", "xml": "<!-- AND：多组条件全部满足才走 true 支路 -->\n<node id=\"160001\" type=\"AND\" title=\"全部满足?\">\n  <branch>\n    <cond op=\"1\" left=\"110001.score\" right=\"0.8\"/>\n    <cond op=\"1\" left=\"110001.hits\" right=\"3\"/>\n  </branch>\n  <out name=\"result\" type=\"boolean\"/>\n  <out name=\"results\" type=\"list<boolean>\"/>\n</node>"}

File: src/assets/nodes_builtin/and.js
// AND 逻辑节点插件（type AND）：多组条件全部满足 → result=true。
// 条件编辑复用 selector 的 branches 组件（props 面板）；画布节点下方渲染各组条件摘要行。
EdFW.register({
  type: "AND", label: "AND", icon: "⛓", category: "logic",
  section: "条件组（全部满足 → true）",
  defaults: {nodeMeta: {title: "AND"},
             inputs: {inputParameters: [],
                      branches: [{condition: {logic: 2, conditions: []}}]},
             outputs: [{name: "result", type: "boolean", description: "全部条件组满足时为 true", fixed: true},
                       {name: "results", type: "list", itemType: "boolean", description: "每组条件的判定结果", fixed: true}]},
  params: [
    { key: "branches", label: "条件组", widget: "branches",
      get(n) { return String((n.data.inputs?.branches || []).length); }, set() {} },
  ],
  nodeH(n) {
    const brs = n.data.inputs?.branches || [];
    return Math.max(1, brs.length) * 16 + 6;   // 每组一行摘要
  },
  body(n, g) {
    // 各组条件摘要（左侧逻辑标记 && ：组内条件表达式拼接，未设置的条件显示 …）
    const brs = n.data.inputs?.branches || [];
    const yBase = HDR_H + Math.max(nodeInputs(n).length, nodeOutputs(n).length) * ROW_H + 4;
    const brs2 = brs.length ? brs : [null];
    brs2.forEach((br, i) => {
      const conds = (br?.condition?.conditions) || [];
      const parts = conds.map(c => {
        const l = c.left?.input?.value?.content;
        const lname = l ? (l.blockID ? `${l.blockID}.${l.name || ''}` : String(l.content ?? '')) : '…';
        const op = (OPERATORS.find(o => o.v === c.operator) || {}).t || '?';
        const r = c.right?.input?.value?.content;
        const rname = r ? (r.blockID ? `${r.blockID}.${r.name || ''}` : String(r.content ?? '')) : '';
        return `${lname} ${op} ${rname}`;
      });
      const txt = parts.length ? parts.join(br?.condition?.logic === 1 ? ' && ' : ' || ') : '（空）';
      const t = document.createElementNS(NS, 'text');
      t.setAttribute('x', 14); t.setAttribute('y', yBase + 11 + i * 16);
      t.setAttribute('class', 'node-field-name');
      t.setAttribute('fill', '#8fa3c8'); t.setAttribute('font-size', '9');
      t.textContent = (i > 0 ? '&& ' : '') + txt.slice(0, 60);
      g.appendChild(t);
    });
  },
});
